segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/AppendProcessor.java

```diff
@@ -255,9 +255,6 @@ private void handleAppendResult(final Append append, Throwable exception) {
             long previousEventNumber;
             synchronized (lock) {
                 previousEventNumber = latestEventNumbers.get(Pair.of(append.getSegment(), append.getWriterId()));
-                Preconditions.checkState(outstandingAppend == append,
```

Comment:
This whole block needs to be moved to the bottom. The lastEventNumbers and waitingAppends need to be updated under the same lock. (it shouldn't be released).

I have a version of this locally, and a unit test, but it is not passing. At the moment I suspect it may be a mockito bug. I'll look into it more. But below is my test in case you want to try it out:
```
@Test(timeout=7000)
    public void testAcksInOrder() {
        String streamSegmentName = "testAppendSegment";
        UUID clientId = UUID.randomUUID();
        ReusableLatch blocker = new ReusableLatch(false);
        byte[] data = new byte[] { 1, 2, 3, 4, 6, 7, 8, 9 };
        StreamSegmentStore store = mock(StreamSegmentStore.class);
        ServerConnection connection = mock(ServerConnection.class);
        InOrder order = Mockito.inOrder(store, connection);
        Mockito.doAnswer(new Answer<Void>() {
            @Override
            public Void answer(InvocationOnMock invocation) throws Throwable {
                blocker.await();
                return null;
            }
        }).when(connection).send(any(DataAppended.class));
        AppendProcessor processor = new AppendProcessor(store, connection, new FailingRequestProcessor());

        setupGetStreamSegmentInfo(streamSegmentName, clientId, store);
        CompletableFuture<Void> result = new CompletableFuture<>();
        when(store.append(streamSegmentName, data, updateEventNumber(clientId, data.length),
                          AppendProcessor.TIMEOUT)).thenReturn(result);
        when(store.append(streamSegmentName, data, updateEventNumber(clientId, 2 * data.length, data.length, 1),
                          AppendProcessor.TIMEOUT)).thenReturn(CompletableFuture.completedFuture(null));

        processor.setupAppend(new SetupAppend(1, clientId, streamSegmentName));
        order.verify(store).getStreamSegmentInfo(anyString(), eq(true), eq(AppendProcessor.TIMEOUT));
        order.verify(connection).send(new AppendSetup(1, streamSegmentName, clientId, 0));
        order.verify(connection, atLeast(0)).resumeReading();
        processor.append(new Append(streamSegmentName, clientId, data.length, Unpooled.wrappedBuffer(data), null));
        order.verify(store).append(streamSegmentName,
                                   data,
                                   updateEventNumber(clientId, data.length),
                                   AppendProcessor.TIMEOUT);
        order.verifyNoMoreInteractions();
        Async.testBlocking(() -> {
            order.verifyNoMoreInteractions();
            result.complete(null);
        }, () -> {
            processor.append(new Append(streamSegmentName, clientId, 2*data.length, Unpooled.wrappedBuffer(data), null));
            order.verify(connection, atLeast(0)).resumeReading();
            order.verifyNoMoreInteractions();
            blocker.release();
        });
        order.verify(connection).send(new DataAppended(clientId, data.length));        
        order.verify(store).append(streamSegmentName,
                                   data,
                                   updateEventNumber(clientId, 2*data.length, data.length, 1),
                                   AppendProcessor.TIMEOUT);
        order.verify(connection).send(new DataAppended(clientId, 2*data.length));
        order.verify(connection, atLeast(0)).resumeReading();
        order.verifyNoMoreInteractions();
    }
```

Reply:
Done, updated the test case too to verify correct handling post the delayed DataAppended ack is sent.